test/e2e/nodenetworkpolicy_test.go

```diff
@@ -39,22 +39,42 @@ func initializeAntreaNodeNetworkPolicy(t *testing.T, data *TestData, toHostNetwo
 	podsPerNamespace = []string{"a"}
 	suffix := randName("")
 	namespaces = make(map[string]TestNamespaceMeta)
-	for _, ns := range []string{"x", "y", "z"} {
-		namespaces[ns] = TestNamespaceMeta{
-			Name: ns + "-" + suffix,
-		}
-	}
 	nodes = make(map[string]string)
-	nodes["x"] = controlPlaneNodeName()
-	nodes["y"] = workerNodeName(1)
 	hostNetworks := make(map[string]bool)
+
+	// Deploy a hostNetwork Pod in Namespace with prefix "x-" on a Node.
+	nodes["x"] = nodeName(0)
 	hostNetworks["x"] = true
-	if toHostNetworkPod {
+	namespaces["x"] = TestNamespaceMeta{
+		Name: "x-" + suffix,
+	}
+
+	if testNodeToNode {
+		// To test NodeNetworkPolicy between Nodes, deploy another hostNetwork Pod in Namespace prefixed with "y-" on
+		// another Node. Pod in Namespace with prefix "z-" is not needed.
+		nodes["y"] = nodeName(1)
 		hostNetworks["y"] = true
+		namespaces["y"] = TestNamespaceMeta{
+			Name: "y-" + suffix,
+		}
 	} else {
+		// To test NodeNetworkPolicy between Node and Pods, deploy another two non-hostNetwork Pods in Namespaces
+		// prefixed with "y-" and "z-", respectively, on another Node.
+		// It is important to note that we avoid deploying non-hostNetwork Pods and hostNetwork Pods on the same Node
+		// for this test. If so, after all test Pods are created, the UDP probing from a non-hostNetwork Pod to the
+		// hostNetwork Pod deployed on the same Node will get a failure. The reason is that the reply packets use the
+		// local Antrea gateway IP as source IP, instead of the local Node IP, which is the destination IP of the request
+		// packets.
```

Comment:
I must be missing something obvious but:
1) why is this specific to UDP?
2) what do you mean by the IP addresses do not match? is there some NAT happening?

Reply:
This is not a new issue. I found it when I added the e2e tests for NodeNetworkPolicy. For example, we have the following pods to test NodeNetworkPolicy:

```
x-d8ummwpo    x-d8ummwpoa-69dd5d9b7b-fzrl2                     8/8   Running   0              26s   192.168.77.100   k8s-node-control-plane   <none>   <none>
y-d8ummwpo    y-d8ummwpoa-5b9ff64477-6vx5w                     8/8   Running   0              26s   192.168.77.101   k8s-node-worker-1        <none>   <none>
z-d8ummwpo    z-d8ummwpoa-9ccbf8f85-dm5gd                      8/8   Running   0              26s   10.10.0.110      k8s-node-control-plane   <none>   <none>
```

We will get the follow error:

```
I0407 02:29:12.333689 2565394 k8s_util.go:229] z-d8ummwpo/a -> x-d8ummwpo/a: expected Con but got Drp: err - <nil> /// stdout -  /// stderr - 1: TIMEOUT: read udp 10.10.0.110:50598->192.168.77.100:80: i/o timeout
2: TIMEOUT: read udp 10.10.0.110:34404->192.168.77.100:80: i/o timeout
3: TIMEOUT: read udp 10.10.0.110:50580->192.168.77.100:80: i/o timeout
FINISHED
```

I run the probe command `for i in $(seq 1 3); do echo -n "${i}: " >&2 && /agnhost connect 192.168.77.100:80 --timeout=1s --protocol=udp && echo "CONNECTED" >&2; done; echo "FINISHED" >&2` on `z` Pod manually. At the same time, I started the tcpdump on antrea-gw0 with this command `tcpdump -i antrea-gw0 port 80 and udp -nve`, and I got this output:

```
02:36:00.391975 a2:ea:32:d3:0f:5c > ca:8a:4a:a2:e0:59, ethertype IPv4 (0x0800), length 51: (tos 0x0, ttl 64, id 55309, offset 0, flags [DF], proto UDP (17), length 37)
    10.10.0.110.36772 > 192.168.77.100.80: UDP, length 9
02:36:00.392178 ca:8a:4a:a2:e0:59 > a2:ea:32:d3:0f:5c, ethertype IPv4 (0x0800), length 64: (tos 0x0, ttl 64, id 38000, offset 0, flags [DF], proto UDP (17), length 50)
    10.10.0.1.80 > 10.10.0.110.36772: UDP, length 22
02:36:01.409093 a2:ea:32:d3:0f:5c > ca:8a:4a:a2:e0:59, ethertype IPv4 (0x0800), length 51: (tos 0x0, ttl 64, id 55327, offset 0, flags [DF], proto UDP (17), length 37)
    10.10.0.110.40658 > 192.168.77.100.80: UDP, length 9
02:36:01.409259 ca:8a:4a:a2:e0:59 > a2:ea:32:d3:0f:5c, ethertype IPv4 (0x0800), length 64: (tos 0x0, ttl 64, id 38027, offset 0, flags [DF], proto UDP (17), length 50)
    10.10.0.1.80 > 10.10.0.110.40658: UDP, length 22
02:36:02.426697 a2:ea:32:d3:0f:5c > ca:8a:4a:a2:e0:59, ethertype IPv4 (0x0800), length 51: (tos 0x0, ttl 64, id 55354, offset 0, flags [DF], proto UDP (17), length 37)
    10.10.0.110.52438 > 192.168.77.100.80: UDP, length 9
02:36:02.426934 ca:8a:4a:a2:e0:59 > a2:ea:32:d3:0f:5c, ethertype IPv4 (0x0800), length 64: (tos 0x0, ttl 64, id 38161, offset 0, flags [DF], proto UDP (17), length 50)
    10.10.0.1.80 > 10.10.0.110.52438: UDP, length 22
```

We can see that there is something wrong with the reply packets.

Reply:
I found that same error on this github action: https://github.com/antrea-io/antrea/actions/runs/8545591837/job/23414615861 started on L3303

Reply:
I see, that makes sense now. I think we should be clearer about the root cause though:
- `agnhost server-hostname` is creating a UDP server that listens on all IP addresses: https://github.com/kubernetes/kubernetes/blob/be4b7176dc131ea842cab6882cd4a06dbfeed12a/test/images/agnhost/serve-hostname/serve_hostname.go#L87
- UDP being connectionless, the reply traffic will "choose" a source IP address based on the routing decision / outgoing interface. This affects UDP, but not TCP.

To workaround the issue, would it be simpler to use `agnhost netexec` instead of `agnhost serve-hostname`? `agnhost netexec` supports `-udp-listen-addresses` so we can provide a specific address for the UDP server to listen to. AFAIK, that would resolve the issue as it will force the reply traffic to use the specified source IP: https://github.com/kubernetes/kubernetes/blob/be4b7176dc131ea842cab6882cd4a06dbfeed12a/test/images/agnhost/netexec/netexec.go#L143

Alternatively, we could use the gateway IP as the destination IP in the test request, as NodeNetworkPolicies should also apply to the gateway? That's not as portable though because in some cases the gateway has no IP (networkPolicyOnly mode).

Reply:
I have tried the following agnhost server Pod entrypoints:

```
 /agnhost netexec --http-port=80 --sctp-port 80 --udp-port 80 --udp-listen-addresses 192.168.77.100,fd00::192:168:77:100
```

The probing of TCP or UDP is okay with the above entrypoint, but for SCTP, I got this from the server side:

```
2024/04/11 13:36:20 Error occurred: failed to read from SCTP client 10.10.1.1/10.176.25.29/192.168.77.101/172.17.0.1/172.18.0.1:43447:EOF
```

And this from client side:

```
bash-5.0#  for i in $(seq 1 3); do echo -n "${i}: " >&2 && /agnhost connect 10.10.0.16:8088 --timeout=1s --protocol=sctp && echo "CONNECTED" >&2; done; echo "FINISHED" >&2
1: CONNECTED
2: OTHER: connection refused
3: OTHER: connection refused
FINISHED
```


```
/agnhost netexec --http-port=0 --udp-port 80 --udp-listen-addresses 192.168.77.100,,fd00::192:168:77:100 & /agnhost serve-hostname --tcp --http=false --port=80 & /agnhost porter
```

The probing of TCP or UDP is also okay with the above entrypoint, but I got a TIMEOUT for SCTP probing.

Reply:
Maybe we could skip the process of probing when the test agnhost Pods are ready for NodeNetworkPolicy e2e tests. Or like before, we schedule the test hostNetwork Pod on Node 0, and non-hostNetwork Pods on Node 1.

Reply:
The first result (` /agnhost netexec`) is expected based on the code (unfortunately) and I can reproduce locally. `nextexec` will kill the server right away after the first `connect`, as it expects the client to send some actual data.

The second result (`/agnhost porter`) is not expected and I cannot reproduce the issue locally. You did set the `SERVE_SCTP_PORT_<port>` environment variable correctly for `porter` right?

```
$ docker run -ti -e SERVE_SCTP_PORT_80=1 registry.k8s.io/e2e-test-images/agnhost:2.29 porter

# 172.17.0.2 is the container's IP address
$ docker run -ti registry.k8s.io/e2e-test-images/agnhost:2.29 connect 172.17.0.2:80 --timeout=1s --protocol=sctp && echo "YES"
YES
$ docker run -ti registry.k8s.io/e2e-test-images/agnhost:2.29 connect 172.17.0.2:80 --timeout=1s --protocol=sctp && echo "YES"
YES
```

Assuming everything is configured correctly, did you see anything strange on tcpdump?

Reply:
I tested the example you gave with docker, and the sctp probing can work. 

Then I tested the case that a hostNetwork agnhost Pod as server and a non-hostNetwork agnhost Pod as client. I found something similar to the result of UDP probing issue and I captured the packets. 

<img width="1207" alt="image" src="https://github.com/antrea-io/antrea/assets/75655411/047608a1-d9fc-4921-ac12-9286c0be58c0">

I tried to find out that if the SCTP can just "listen" on a specific address. Unfortunately, I didn't find the related code in the link https://github.com/kubernetes/kubernetes/blob/be4b7176dc131ea842cab6882cd4a06dbfeed12a/test/images/agnhost/netexec/netexec.go.

Reply:
That's interesting. I thought that because SCTP is connection oriented (like TCP) it would not suffer from the same issue as UDP. But I don't know much about SCTP and I may be missing something. Anyway, while the behavior is surprising, we don't need to spend more time on this. Thanks for taking the time to test it out as you have.